### PR TITLE
Aded catch in R^2 calculation for case with few samples

### LINKED ISSUE #5306: Index out of range exception in execute
### System information

windows 10 version 1803
NET Version (3.1.300): 
AutoML version : 0.17.1
(also tested on version 0.14.0)
### Issue

I am trying to use AutoML for the first time. but I get IndexOutOfRangeException on execute method call.

### Source code / logs
static void Main(string[] args)
		{
			MLContext mlContext = new MLContext();
			IDataView trainDataView = mlContext.Data.LoadFromTextFile<ModelInput>("input.csv", hasHeader: true);
			var experimentSettings = new RegressionExperimentSettings();
			experimentSettings.MaxExperimentTimeInSeconds = 100;
			RegressionExperiment experiment = mlContext.Auto().CreateRegressionExperiment(experimentSettings);
			ExperimentResult<RegressionMetrics> experimentResult = experiment.Execute(trainDataView, "Withdrawal");
			RegressionMetrics metrics = experimentResult.BestRun.ValidationMetrics;
			Console.WriteLine($"R-Squared: {metrics.RSquared:0.##}");
			Console.WriteLine($"Root Mean Squared Error: {metrics.RootMeanSquaredError:0.##}");
		}

Error:

Unhandled exception. System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at Microsoft.ML.AutoML.CrossValSummaryRunner`1.Run(SuggestedPipeline pipeline, DirectoryInfo modelDirectory, Int32 iterationNum)
   at Microsoft.ML.AutoML.Experiment`2.Execute()
   at Microsoft.ML.AutoML.ExperimentBase`2.Execute(ColumnInformation columnInfo, DatasetColumnInfo[] columns, IEstimator`1 preFeaturizer, IProgress`1 progressHandler, IRunner`1 runner)
   at Microsoft.ML.AutoML.ExperimentBase`2.ExecuteCrossValSummary(IDataView[] trainDatasets, ColumnInformation columnInfo, IDataView[] validationDatasets, IEstimator`1 preFeaturizer, IProgress`1 progressHandler)
   at Microsoft.ML.AutoML.ExperimentBase`2.Execute(IDataView trainData, ColumnInformation columnInformation, IEstimator`1 preFeaturizer, IProgress`1 progressHandler)
   at Microsoft.ML.AutoML.ExperimentBase`2.Execute(IDataView trainData, String labelColumnName, String samplingKeyColumn, IEstimator`1 preFeaturizer, IProgress`1 progressHandler)
   at FunWithMlnet.Program.Main(String[] args) in C:\Users\User\source\repos\FunWithWithdrawalsData\FunWithMlnet\Program.cs:line 18

### PR DESCRIPTION
**Pull request by mstfbl**
This PR fixes #5306 by adding a catch for the calculation of `R^2` during metric calculation. When there is less than two rows of data used for the calculation of `R^2`, the returned value becomes `-Infinity`, whereas it should be returning `NaN`.

An example of this behavior in another ML framework is in scikit-learn, in the following lines:

https://github.com/scikit-learn/scikit-learn/blob/fd237278e895b42abe8d8d09105cbb82dc2cbba7/sklearn/metrics/_regression.py#L587-L590

### CODE REVIEW

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs [hunk 1]
```diff
@@ -102,6 +102,10 @@ public override double RSquared
                 {
                     get
                     {
+                        // RSquared value cannot be well-defined with less than two samples.
+                        // Return NaN instead of -Infinity.
+                        if ((TotalLabelSquaredW - TotalLabelW * TotalLabelW / SumWeights) == 0)
+                            return double.NaN;
```
Comment (Lynx1820) [inline src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:108]: Could you add a test for this?

Comment (mstfbl) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:108]: There isn't really a direct way of testing this apart from forcing a cross validation scenario in AutoML where the size of the folds in cross validation is less than 2. If a test for this is necessary then I will go this route.

Comment (harishsk) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:108]: Why is AutoML required for this? 

---
In reply to: [459046236](https://github.com/dotnet/machinelearning/pull/5319#discussion_r459046236) [](ancestors = 459046236)

Comment (mstfbl) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:108]: My bad on wording, AutoML isn't required to test this, the original issue exposed this bug through an AutoML experiment. A call to `mlContext.Regression.Evaluate(prediction)` where `prediction` is a one-row IDataView would be enough to test this.

Comment (justinormont) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:108]: Could make a test which trains a regression model then tests on one row of data, checking that the output is `NaN` for R2.

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs [hunk 2]
```diff
@@ -102,7 +102,16 @@ public override double RSquared
...
-                        return SumWeights > 0 ? 1 - TotalL2Loss / (TotalLabelSquaredW - TotalLabelW * TotalLabelW / SumWeights) : 0;
+                        // RSquared value cannot be well-defined with less than two samples.
+                        // Return NaN instead of -Infinity.
+                        if (SumWeights > 0)
+                        {
+                            if ((TotalLabelSquaredW - TotalLabelW * TotalLabelW / SumWeights) == 0)
+                                return double.NaN;
+                            return 1 - TotalL2Loss / (TotalLabelSquaredW - TotalLabelW * TotalLabelW / SumWeights);
+                        }
+                        else
```
Comment (Lynx1820) [inline src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:113]: nit: Can we get rid of the ` else` and just return 0?

File: test/Microsoft.ML.Functional.Tests/Validation.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@
 using Microsoft.ML.Trainers.LightGbm;
 using Xunit;
 using Xunit.Abstractions;
+using static Microsoft.ML.TrainCatalogBase;
```
Comment (Lynx1820) [inline test/Microsoft.ML.Functional.Tests/Validation.cs:15]: What is this used for?

Comment (mstfbl) [reply test/Microsoft.ML.Functional.Tests/Validation.cs:15]: It's to call the `mlContext.Regression.CrossValidate` function on line 160.

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs (hunk 2 above)
Comment (mstfbl) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:113]: Fixed!

File: test/Microsoft.ML.Functional.Tests/Validation.cs [hunk 4]
```diff
@@ -138,5 +139,27 @@ public void TrainWithValidationSet()
...
+            var mlContext = new MLContext(1);
+            // Get data and set up sample regression pipeline.
+            var data = mlContext.Data.LoadFromTextFile<Iris>(TestCommon.GetDataPath(DataDir, TestDatasets.iris.trainFilename), hasHeader: true);
+            var dataFirstTenRows = mlContext.Data.TakeRows(data, 10);
+            var pipeline = mlContext.Transforms.Concatenate("Features", Iris.Features)
+                .Append(mlContext.Regression.Trainers.OnlineGradientDescent());
+
+            // Check that NaN is returned with fold given less than 2 rows of training data.
+            // With dataset of 10 rows, number of folds will be 6.
+            var cvResults = mlContext.Regression.CrossValidate(dataFirstTenRows, pipeline, numberOfFolds: 6);
```
Comment (justinormont) [inline test/Microsoft.ML.Functional.Tests/Validation.cs:160]: Could use a normal training, then evaluate on one row of data. This way you could avoid the randomness of creating the CV folds (which is current curtailed using a seed).

One secondary test would be testing zero rows in the scoring dataset.

Comment (mstfbl) [reply test/Microsoft.ML.Functional.Tests/Validation.cs]: Edited unit test per the above, and added secondary test for the 0 row scenario. Thanks!

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs (hunk 1 above)
Comment (mstfbl) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs]: Added test and completed the request change to merge PR.
